CRM/Mailing/BAO/MailingJob.php

```diff
@@ -705,20 +707,23 @@ public function writeToDB(
 
       //check whether activity is already created for this mailing.
       //if yes then create only target contact record.
-      $query = "
+      if (!$activityID) {
```

Comment:
Picking up @ufundo's feedback my thinking would be to do something like here

if (!isset(Civi::$statics[__CLASS__]['activity_ids'])) {
 Civi::$statics[__CLASS__]['activity_ids'] = [];
}
if (!isset(Civi::$statics[__CLASS__]['activity_ids'][$this->mailing_id])) {
  // run the query
  and set Civi::$statics[__CLASS__]['activity_ids'][$this->mailing_id] = found activity id and then just use Civi::$statics[__CLASS__]['activity_ids'][$this->mailing_id] as the reference to the activity id going forward or something